internal/api/mfa.go

```diff
@@ -205,6 +364,115 @@ func (a *API) ChallengeFactor(w http.ResponseWriter, r *http.Request) error {
 	})
 }
 
+func (a *API) verifySMSFactor(w http.ResponseWriter, r *http.Request, user *models.User, factor *models.Factor, params *VerifyFactorParams) error {
+	ctx := r.Context()
+	config := a.config
+	db := a.db.WithContext(ctx)
+	if !config.MFA.SMS.VerifyEnabled {
+		// TODO: Replace with appropriate factor type
+		return unprocessableEntityError(ErrorCodeMFAVerifyDisabled, "mfa verification is disabled for sms factors")
+	}
+	currentIP := utilities.GetIPAddress(r)
+
+	if !factor.IsOwnedBy(user) {
+		return internalServerError(InvalidFactorOwnerErrorMessage)
+	}
+
+	challenge, err := models.FindChallengeByID(db, params.ChallengeID)
+	if err != nil && models.IsNotFoundError(err) {
+		return notFoundError(ErrorCodeMFAFactorNotFound, "MFA factor with the provided challenge ID not found")
+	} else if err != nil {
+		return internalServerError("Database error finding Challenge").WithInternalError(err)
+	}
+
+	if challenge.VerifiedAt != nil || challenge.IPAddress != currentIP {
+		return unprocessableEntityError(ErrorCodeMFAIPAddressMismatch, "Challenge and verify IP addresses mismatch")
+	}
+
+	if challenge.HasExpired(config.MFA.ChallengeExpiryDuration) {
+		if err := db.Destroy(challenge); err != nil {
+			return internalServerError("Database error deleting challenge").WithInternalError(err)
+		}
+		return unprocessableEntityError(ErrorCodeMFAChallengeExpired, "MFA challenge %v has expired, verify against another challenge or create a new challenge.", challenge.ID)
+	}
+	valid := subtle.ConstantTimeCompare([]byte(challenge.OtpCode), []byte(params.Code)) == 1
```

Comment:
if hashing in the the OTP in the DB we'll need 
```suggestion
	valid := subtle.ConstantTimeCompare([]byte(challenge.OtpCode), sha256.Sum256([]byte(params.Code))) == 1
```

Reply:
Somewhat against hashing for SMS as it makes implementing a "resend" function hard -- it will not resend, but rather send a new code, which can lead to user confusion when 3 SMS messages are received at once.

In this case the OtpCode must be encrypted.

Reply:
happy with encryption over hashing given that additional context

Reply:
Sounds good, let's encrypt.